dstopic/packages/parse-package-arg/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+use lazy_static::lazy_static;
+use regex::Regex;
+use semver::{Version, VersionReq};
+use thiserror::Error;
+use url::{Host, Url};
+
+#[derive(Debug, Error)]
+pub enum PackageArgError {
+    #[error("Failed to parse package arg.")]
+    ParseError,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PackageArg {
+    Dir {
+        name: Option<String>,
+        path: PathBuf,
+    },
+    Alias {
+        name: String,
+        package: Box<PackageArg>,
+    },
+    Tag {
+        host: Option<Host>,
+        name: String,
+        tag: String,
+    },
+    Version {
+        host: Option<Host>,
+        name: String,
+        version: Version,
+    },
+    Range {
+        host: Option<Host>,
+        name: String,
+        range: VersionReq,
+    },
+}
+
+lazy_static! {
+    static ref SPLITTER: Regex =
+        Regex::new(r"(?i)^(?P<name>(?:[^@]+/legacy/@[^@]+|[^@]+))(?:@(?P<spec>.*))?$").unwrap();
+    static ref PKG: Regex = Regex::new(r"^(?P<host>[^/]+/)(?P<name>[^/]+/[^/]+)$").unwrap();
+    static ref LEGACY: Regex =
+        Regex::new(r"(?i)^(?P<host>[^/]+/)(?P<name>legacy/[^/]+(?:/[^/]+)?)$").unwrap();
+    static ref IS_FILE: Regex = Regex::new(r"(?i)^(?:(?:[a-z]:)?[/\\]|\.[/\\]|file:)").unwrap();
+}
+
+impl PackageArg {
+    pub fn from_string<S: AsRef<str>>(s: S) -> Result<PackageArg> {
+        let s: String = s.as_ref().into();
+        let matches = SPLITTER
+            .captures(&s)
+            .ok_or_else(|| PackageArgError::ParseError)
+            .with_context(|| format!("Invalid package arg: {}", s))?;
+        let name = matches.name("name").map(|name| name.as_str().to_owned());
+        let spec = matches.name("spec").map(|name| name.as_str().to_owned());
+        if let Some(name) = name {
+            if IS_FILE.is_match(&name) {
+                Self::resolve(None, Some(name))
+            } else {
+                Self::resolve(Some(name), spec)
+            }
+        } else {
+            Self::resolve(name, spec)
+        }
+    }
+
+    pub fn resolve(name: Option<String>, spec: Option<String>) -> Result<PackageArg> {
+        if let Some(spec) = spec {
+            if IS_FILE.is_match(&spec) {
+                from_dir(name, spec)
+            } else if name.is_none() {
+                Err(PackageArgError::ParseError).with_context(|| {
+                    format!(
+                        "Tried to resolve a registry spec ({}) without a name.",
+                        spec
+                    )
+                })?
+            } else if spec.starts_with("pkg:") {
+                from_alias(name.unwrap(), spec)
+            } else {
+                from_registry(name.unwrap(), Some(spec))
+            }
+        } else if let Some(name) = name {
+            from_registry(name, None)
+        } else {
+            Err(PackageArgError::ParseError).with_context(|| format!("I don't know"))?
```

Comment:
& this is the null case, where someone said `resolve(None, None)`.

Reply:
oooh, yes. I should probably have a better error message here.